### ISSUE TITLE
"delete" issue

### ISSUE THREAD

**Issue by nikita-volkov** [nikita-volkov/focus#7]
It does not seem to be fixed in 1.0.1.4 (either in focus or stm-containers / hamt). The code below loops infinitely:
```haskell
#!/usr/bin/env stack
{- stack
  script
  --resolver lts-16.26
  --package stm
  --package async
  --package focus
  --package transformers
  --package hashable
  --package unsafe
  --package containers
  --package nonempty-containers
  --package stm-containers
  --extra-dep stm-hamt-1.2.0.4@sha256:7957497c022554b7599e790696d1a3e56359ad99e5da36a251894c626ca1f60a,3970
  --extra-dep primitive-extras-0.8@sha256:fca0310150496867f5b9421fe1541ecda87fae17eae44885a29f9c52dd00c8ff,2963
  --extra-dep primitive-unlifted-0.1.3.0@sha256:a98f827740f5dcf097d885b3a47c32f4462204449620abc9d51b8c4f8619f9e6,1427
-}

import           Control.Concurrent
import           Control.Concurrent.STM
import           Control.Monad
import           Data.Hashable          (Hashable)
import           Data.Maybe
import qualified Data.Sequence          as S
import           Data.Sequence.NonEmpty (NESeq, (|>))
import qualified Data.Sequence.NonEmpty as NES
import qualified Focus                  as F
import qualified StmContainers.Map      as SM
import           System.IO.Unsafe
import           System.Unsafe

popQueue :: (Hashable k, Eq k) => k -> SM.Map k (NESeq v) -> IO (Maybe v)
popQueue key queue = atomically $ do
  element'm <- SM.focus popQueueF key queue
  when (isNothing element'm) $ SM.delete key queue -- this doesn't delete either!!!
  pure element'm
  where
    popQueueF = do
      element'm <- fmap (fmap NES.head) F.lookup
      queue'm <- fmap (join . (fmap (NES.nonEmptySeq . NES.tail))) F.lookup
      unsafePerformIO $ do -- debugging
        let elem_ = coerce element'm :: Maybe (String)
        let elem_t = coerce queue'm :: Maybe (NESeq String)
        putStrLn "***"
        putStrLn $ show $ elem_
        putStrLn $ show $ elem_t
        putStrLn "***"
        threadDelay 100000
        pure F.lookup
      -- F.alter (join . fmap (NES.nonEmptySeq . NES.tail))
      case queue'm of
        Just queue_ -> F.insert queue_
        Nothing     -> F.delete -- doesn't delete!!!
      pure element'm

pushQueue :: (Hashable k, Eq k) => k -> v -> SM.Map k (NESeq v) -> IO ()
pushQueue key value = atomically . SM.focus pushQueueF key
  where
    pushQueueF = F.insertOrMerge (<>) (NES.singleton value)


main :: IO ()
main = do
  putStrLn "\n"
  queue <- SM.newIO :: IO (SM.Map Int (NESeq String))
  pushQueue 1 "foo" queue
  pushQueue 1 "bar" queue
  pushQueue 2 "fooz" queue
  pushQueue 2 "baz" queue
  putStrLn $ "DRAIN-----1"
  drainQ 1 queue putStrLn
  putStrLn $ "DRAIN-----2"
  drainQ 2 queue putStrLn
    where
      drainQ :: (Hashable k, Eq k) => k -> SM.Map k (NESeq v) -> (v -> IO a) -> IO ()
      drainQ key queue act = do
        element'm <- popQueue key queue
        case element'm of
          Just element -> do act element
                             drainQ key queue act
          Nothing -> pure ()
```

_Originally posted by @carbolymer in https://github.com/nikita-volkov/focus/issues/5#issuecomment-744010388_

**Comment by nikita-volkov** [comment]
@carbolymer

Since it's not related to the issue in which you've posted I've created this one as a dedicate issue.

Can you provide a reduced reproduction? Too many things happening in this code to spot an issue.

**Comment by carbolymer** [comment]
Here:
```haskell
#!/usr/bin/env stack
{- stack
  script
  --resolver lts-16.26
  --package stm
  --package focus
  --package hashable
  --package unsafe
  --package stm-containers
  --extra-dep stm-hamt-1.2.0.4@sha256:7957497c022554b7599e790696d1a3e56359ad99e5da36a251894c626ca1f60a,3970
  --extra-dep primitive-extras-0.8@sha256:fca0310150496867f5b9421fe1541ecda87fae17eae44885a29f9c52dd00c8ff,2963
  --extra-dep primitive-unlifted-0.1.3.0@sha256:a98f827740f5dcf097d885b3a47c32f4462204449620abc9d51b8c4f8619f9e6,1427
-}

import           Control.Concurrent
import           Control.Concurrent.STM
import           Control.Monad
import           Data.Hashable          (Hashable)
import           Data.Maybe
import qualified Focus                  as F
import qualified StmContainers.Map      as SM
import           System.IO.Unsafe
import           System.Unsafe

delMap :: (Hashable k, Eq k) => k -> SM.Map k v -> IO (Maybe v)
delMap key stmMap = atomically $ SM.focus delMapF key stmMap
  where
    -- this is just `F.lookup <* F.delete` in a more verbose form
    delMapF = do
      element'm <- F.lookup
      -- unsafePerformIO $ do -- uncomment this block for printing what element is being processed
      --   let elem_ = coerce element'm :: Maybe (String)
      --   putStrLn "***"
      --   putStrLn $ show $ elem_
      --   putStrLn "***"
      --   threadDelay 100000
      --   pure F.lookup

      F.delete -- this doesn't do anything!
      pure element'm

insertMap :: (Hashable k, Eq k) => k -> v -> SM.Map k v -> IO ()
insertMap key value = atomically . SM.insert value key

main :: IO ()
main = do
  putStrLn "\n"
  stmMap <- SM.newIO :: IO (SM.Map Int String)
  insertMap 1 "foo" stmMap
  insertMap 2 "bar" stmMap
  putStrLn $ "POP MAP 1"
  delMap 1 stmMap >>= (putStrLn . show) -- this returns Just "foo"
  delMap 1 stmMap >>= (putStrLn . show) -- this returns Just "foo" -- INVALID - SHOULD RETURN Nothing
  delMap 1 stmMap >>= (putStrLn . show) -- this returns Just "foo" -- INVALID - SHOULD RETURN Nothing
  putStrLn $ "POP MAP 2"
  delMap 2 stmMap >>= (putStrLn . show) -- this returns Just "bar"
  delMap 2 stmMap >>= (putStrLn . show) -- this returns Just "bar" -- INVALID - SHOULD RETURN Nothing
  delMap 2 stmMap >>= (putStrLn . show) -- this returns Just "bar" -- INVALID - SHOULD RETURN Nothing

```

The problem lies in `delMap` function. The `delMapF` is just `F.lookup <* F.delete` which does not delete! the output of this script is:
```
POP MAP 1
Just "foo"
Just "foo"
Just "foo"
POP MAP 2
Just "bar"
Just "bar"
Just "bar"
```
where it should be:
```
POP MAP 1
Just "foo"
Nothing
Nothing
POP MAP 2
Just "bar"
Nothing
Nothing
```

So, there's no possibility to delete items from the map at this moment.

**Comment by nikita-volkov** [comment]
Fixed in focus-1.0.2. Thanks for reporting @carbolymer!
